File: src/main/java/xyz/ivyxjc/pubg4j/JsonBuilder.java
package xyz.ivyxjc.pubg4j;

import com.google.gson.Gson;
import com.google.gson.internal.LinkedTreeMap;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.util.Assert;
import xyz.ivyxjc.pubg4j.constants.JsonConstants;
import xyz.ivyxjc.pubg4j.entity.ParticipantDetail;
import xyz.ivyxjc.pubg4j.entity.PubgMatch;
import xyz.ivyxjc.pubg4j.entity.PubgMatchDetail;
import xyz.ivyxjc.pubg4j.entity.PubgPlayer;
import xyz.ivyxjc.pubg4j.entity.Roster;
import xyz.ivyxjc.pubg4j.exception.UnsupportedPubgElementException;
import xyz.ivyxjc.pubg4j.types.GameMode;
import xyz.ivyxjc.pubg4j.types.PlatformRegion;

/**
 * @author Ivyxjc
 * @since 4/22/2018
 */
@Slf4j
@Service
public class JsonBuilder {
    // DateTimeFormatter is immutable and thread-safe.
    private static final DateTimeFormatter dateTimeFormatter =
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'");

    // Gson is thread-safe
    private static final Gson gson = new Gson();

    public PubgPlayer buildPlayer(InputStream in) throws UnsupportedPubgElementException {
        PubgPlayer pubgPlayer = new PubgPlayer();
        Reader jsonReader = new InputStreamReader(in);
        Map map = gson.fromJson(jsonReader, HashMap.class);
        LinkedTreeMap dataMap = null;
        LinkedTreeMap attributeMap = null;
        dataMap = (LinkedTreeMap) ((List) map.get("data")).get(0);
        Assert.notNull(dataMap, "Player's data map should not be null");
        if (dataMap.get("attributes") instanceof LinkedTreeMap) {
            attributeMap = (LinkedTreeMap) dataMap.get("attributes");
        }
        Assert.notNull(attributeMap, "Player's attributea Map should not be null");
        pubgPlayer.setPlayerId((String) dataMap.get("id"));
        pubgPlayer.setCreatedAt(
            LocalDateTime.parse((String) attributeMap.get("createdAt"), dateTimeFormatter));
        pubgPlayer.setName((String) attributeMap.get("name"));
        pubgPlayer.setShardId(PlatformRegion.enumOf((String) attributeMap.get("shardId")));
        pubgPlayer.setTitleId((String) attributeMap.get("titleId"));
        pubgPlayer.setUpdatedAt(
            LocalDateTime.parse((String) attributeMap.get("updatedAt"), dateTimeFormatter));
        List matchList =
            (ArrayList) ((LinkedTreeMap) ((LinkedTreeMap) dataMap.get("relationships")).get(
                "matches")).get("data");
        for (int i = 0; i < matchList.size(); i++) {
            if (matchList.get(i) instanceof LinkedTreeMap) {
                LinkedTreeMap ltm = (LinkedTreeMap) matchList.get(i);
                if (ltm.size() < 2) {
                    continue;
                }
                PubgMatch pubgMatch = new PubgMatch();
                String ltmId = (String) ltm.get("id");
                Assert.notNull(ltmId, "Player's data's match id should not be nul");
                pubgMatch.setMatchId((String) ltm.get("id"));
                pubgPlayer.getMatches().add(pubgMatch);
            }
        }
        return pubgPlayer;
    }

    public PubgMatchDetail buildMatch(InputStream in) throws UnsupportedPubgElementException {
        PubgMatchDetail matchDetail = new PubgMatchDetail();
        HashMap map = gson.fromJson(new InputStreamReader(in), HashMap.class);
        LinkedTreeMap dataMap = (LinkedTreeMap) map.get("data");
        Assert.notNull(dataMap, "dataMap should not be null");
        LinkedTreeMap attributesMap = (LinkedTreeMap) dataMap.get("attributes");
        Assert.notNull(attributesMap, "attributesMap should not be null");
        LinkedTreeMap relationshipsMap = (LinkedTreeMap) dataMap.get("relationships");
        Assert.notNull(relationshipsMap, "relationshipsMap should not be null");
        ArrayList rosterMapList =
            (ArrayList) ((LinkedTreeMap) relationshipsMap.get("rosters")).get("data");
        Assert.notNull(rosterMapList, "rosterMapList should not be null");
        ArrayList includedList = (ArrayList) map.get("included");
        Assert.notNull(includedList, "includedList should not be null");
        matchDetail.setMatchId((String) dataMap.get("id"));
        matchDetail.setCreatedAt(
            LocalDateTime.parse((String) attributesMap.get("createdAt"), dateTimeFormatter));
        matchDetail.setDuration(((Double) attributesMap.get("duration")).intValue());
        matchDetail.setGameMode(GameMode.enumOf((String) attributesMap.get("gameMode")));
        matchDetail.setShardId(PlatformRegion.enumOf((String) attributesMap.get("shardId")));
        matchDetail.setTitleId((String) attributesMap.get("titleId"));
        for (int i = 0; i < rosterMapList.size(); i++) {
            Roster roster = new Roster();
            LinkedTreeMap ltm = (LinkedTreeMap) rosterMapList.get(i);
            Assert.notNull(includedList, "ltm should not be null");
            String tmpId = (String) ltm.get("id");
            Assert.notNull(tmpId, "ltm.get(id) should not be null");
            roster.setRosterId(tmpId);
            matchDetail.getRostetMap().put(roster.getRosterId(), roster);
        }

        for (int i = 0; i < includedList.size(); i++) {
            if (includedList.get(i) instanceof LinkedTreeMap) {
                LinkedTreeMap ltm = (LinkedTreeMap) includedList.get(i);
                if (JsonConstants.TYPE_PARTICIPANT.equals(ltm.get("type"))) {
                    ParticipantDetail participantDetail = new ParticipantDetail();
                    String tmpId = (String) ltm.get("id");
                    Assert.notNull(tmpId, "participant's id should not be null");
                    participantDetail.setParticipantId(tmpId);
                    LinkedTreeMap statsLtm =
                        (LinkedTreeMap) ((LinkedTreeMap) ltm.get("attributes")).get("stats");
                    parseParticipant(participantDetail, statsLtm);
                }
                if (JsonConstants.TYPE_ROSTER.equals(ltm.get("type"))) {
                    String tmpId = (String) ltm.get("id");
                    Assert.notNull(tmpId, "roster's id should not be null");
                    Roster roster = matchDetail.getRostetMap().get(tmpId);
                    LinkedTreeMap atbLtm = (LinkedTreeMap) ltm.get("attributes");
                    LinkedTreeMap statsLtm = (LinkedTreeMap) atbLtm.get("stats");
                    roster.setShardId(PlatformRegion.enumOf((String) atbLtm.get("shardId")));
                    if ("true".equals((String) atbLtm.get("won"))) {
                        roster.setWin(true);
                    } else {
                        roster.setWin(false);
                    }
                    roster.setRank(((Double) statsLtm.get("rank")).intValue());
                    roster.setTeamId(((Double) statsLtm.get("teamId")).intValue());
                }
            }
        }
        return matchDetail;
    }

    private void parseParticipant(ParticipantDetail participant, LinkedTreeMap treeMap) {
        participant.setDbnos(((Double) treeMap.get("DBNOs")).intValue());
        participant.setAssists(((Double) treeMap.get("assists")).intValue());
        participant.setBoosts(((Double) treeMap.get("boosts")).intValue());
        participant.setDamageDealt((Double) treeMap.get("damageDealt"));
        participant.setDeathType((String) treeMap.get("deathType"));
        participant.setHeadshotKills(((Double) treeMap.get("headshotKills")).intValue());
        participant.setHeals(((Double) treeMap.get("heals")).intValue());
        participant.setKillPlace(((Double) treeMap.get("killPlace")).intValue());
        participant.setKillPoints(((Double) treeMap.get("killPoints")).intValue());
        participant.setKillPointsDelta((Double) treeMap.get("killPointsDelta"));
        participant.setKillStreaks(((Double) treeMap.get("killStreaks")).intValue());
        participant.setKills(((Double) treeMap.get("kills")).intValue());
        participant.setLastKillPoints(((Double) treeMap.get("lastKillPoints")).intValue());
        participant.setLastWinPoints(((Double) treeMap.get("lastWinPoints")).intValue());
        participant.setLongestKill((Double) treeMap.get("longestKill"));
        participant.setMostDamage(((Double) treeMap.get("mostDamage")).intValue());
        participant.setName((String) treeMap.get("name"));
        participant.setPlayerId((String) treeMap.get("playerId"));
        participant.setRevives(((Double) treeMap.get("revives")).intValue());
        participant.setRideDistance((Double) treeMap.get("rideDistance"));
        participant.setRoadKills(((Double) treeMap.get("roadKills")).intValue());
        participant.setTeamKills(((Double) treeMap.get("teamKills")).intValue());
        participant.setTimeSurvived(((Double) treeMap.get("timeSurvived")).intValue());
        participant.setVehicleDestroys(((Double) treeMap.get("vehicleDestroys")).intValue());
        participant.setWalkDistance((Double) treeMap.get("walkDistance"));
        participant.setWeaponsAcquired(((Double) treeMap.get("weaponsAcquired")).intValue());
        participant.setWinPlace(((Double) treeMap.get("winPlace")).intValue());
        participant.setWinPoints(((Double) treeMap.get("winPoints")).intValue());
        participant.setWinPointsDelta((Double) treeMap.get("winPointsDelta"));
    }
}

File: src/main/kotlin/xyz/ivyxjc/pubg4j/JsonBuilder.kt
package xyz.ivyxjc.pubg4j

import com.google.gson.Gson
import com.google.gson.internal.LinkedTreeMap
import org.springframework.stereotype.Service
import org.springframework.util.Assert
import xyz.ivyxjc.pubg4j.entity.*
import xyz.ivyxjc.pubg4j.exception.UnsupportedPubgElementException
import xyz.ivyxjc.pubg4j.types.GameMode
import xyz.ivyxjc.pubg4j.types.MapName
import xyz.ivyxjc.pubg4j.types.PlatformRegion
import xyz.ivyxjc.pubg4j.utils.JsonConstants
import java.io.InputStream
import java.io.InputStreamReader
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter


/**
 * @author Ivyxjc
 * @since 4/22/2018
 */
@Service
class JsonBuilder {

    @Throws(UnsupportedPubgElementException::class)
    fun buildPlayer(`in`: InputStream): PubgPlayer {
        val jsonReader = InputStreamReader(`in`)
        val map = gson.fromJson(jsonReader, HashMap::class.java)
        var dataMap: LinkedTreeMap<*, *>?
        var attributeMap: LinkedTreeMap<*, *>? = null
        dataMap = (map["data"] as List<*>)[0] as LinkedTreeMap<*, *>
        Assert.notNull(dataMap, "Player's data map should not be null")
        if (dataMap["attributes"] is LinkedTreeMap<*, *>) {
            attributeMap = dataMap["attributes"] as LinkedTreeMap<*, *>
        }
        Assert.notNull(attributeMap, "Player's attributea Map should not be null")
        val pubgPlayer = PubgPlayer(dataMap["id"] as String)
        pubgPlayer.createdAt = LocalDateTime.parse(attributeMap!!["createdAt"] as String, dateTimeFormatter)
        pubgPlayer.name = attributeMap["name"] as String
        pubgPlayer.shardId = PlatformRegion.enumOf(attributeMap["shardId"] as String)
        pubgPlayer.titleId = attributeMap["titleId"] as String
        pubgPlayer.updatedAt = LocalDateTime.parse(attributeMap["updatedAt"] as String, dateTimeFormatter)
        val matchList = ((dataMap["relationships"] as LinkedTreeMap<*, *>)["matches"] as LinkedTreeMap<*, *>)["data"] as ArrayList<*>
        for (i in matchList.indices) {
            if (matchList[i] is LinkedTreeMap<*, *>) {
                val ltm = matchList[i] as LinkedTreeMap<*, *>
                if (ltm.size < 2) {
                    continue
                }
                val pubgMatch = PubgMatch(ltm["id"] as String, pubgPlayer.playerId, pubgPlayer.shardId!!)
                val ltmId = ltm["id"] as String
                Assert.notNull(ltmId, "Player's data's match id should not be nul")
                pubgPlayer.matches.add(pubgMatch)
            }
        }
        return pubgPlayer
    }

    @Throws(UnsupportedPubgElementException::class)
    fun buildMatch(`in`: InputStream): PubgMatchDetail {
        val map = gson.fromJson(InputStreamReader(`in`), HashMap::class.java)
        val dataMap = map["data"] as LinkedTreeMap<*, *>
        Assert.notNull(dataMap, "dataMap should not be null")
        val attributesMap = dataMap["attributes"] as LinkedTreeMap<*, *>
        Assert.notNull(attributesMap, "attributesMap should not be null")
        val relationshipsMap = dataMap["relationships"] as LinkedTreeMap<*, *>
        Assert.notNull(relationshipsMap, "relationshipsMap should not be null")
        val rosterMapList = (relationshipsMap["rosters"] as LinkedTreeMap<*, *>)["data"] as ArrayList<*>
        Assert.notNull(rosterMapList, "rosterMapList should not be null")
        val includedList = map["included"] as ArrayList<*>
        Assert.notNull(includedList, "includedList should not be null")
        val matchDetail = PubgMatchDetail(dataMap["id"] as String)
        matchDetail.createdAt = LocalDateTime.parse(attributesMap["createdAt"] as String, dateTimeFormatter)
        matchDetail.duration = (attributesMap["duration"] as Double).toInt()
        matchDetail.gameMode = GameMode.enumOf(attributesMap["gameMode"] as String)
        matchDetail.mapName = MapName.enumOf(attributesMap["mapName"] as String)
        matchDetail.shardId = PlatformRegion.enumOf(attributesMap["shardId"] as String)
        matchDetail.titleId = attributesMap["titleId"] as String
        for (i in rosterMapList.indices) {
            val ltm = rosterMapList[i] as LinkedTreeMap<*, *>
            Assert.notNull(includedList, "ltm should not be null")
            val tmpId = ltm["id"] as String
            Assert.notNull(tmpId, "ltm.get(id) should not be null")
            val roster = PubgRoster(tmpId)
            matchDetail.rosterMap.put(roster.rosterId, roster)
        }

        for (i in includedList.indices) {
            if (includedList[i] is LinkedTreeMap<*, *>) {
                val ltm = includedList[i] as LinkedTreeMap<*, *>
                if (JsonConstants.TYPE_PARTICIPANT == ltm["type"]) {
                    val tmpId = ltm["id"] as String
                    Assert.notNull(tmpId, "participant's id should not be null")
                    val participantDetail = matchDetail.participantDetailMap.get(tmpId)
                        ?: PubgParticipant(tmpId)
                    val statsLtm = (ltm["attributes"] as LinkedTreeMap<*, *>)["stats"] as LinkedTreeMap<*, *>
                    parseParticipant(participantDetail, statsLtm)
                    participantDetail.matchId = matchDetail.matchId
                    matchDetail.participantDetailMap
                        .putIfAbsent(participantDetail.participantId, participantDetail)

                }
                if (JsonConstants.TYPE_ROSTER == ltm["type"]) {
                    val tmpRosterId = ltm["id"] as String
                    Assert.notNull(tmpRosterId, "roster's id should not be null")
                    val roster = matchDetail.rosterMap.get(tmpRosterId)
                    val atbLtm = ltm["attributes"] as LinkedTreeMap<*, *>
                    val statsLtm = atbLtm["stats"] as LinkedTreeMap<*, *>
                    roster?.matchId = matchDetail.matchId
                    roster?.shardId = PlatformRegion.enumOf(atbLtm["shardId"] as String)
                    roster?.win = "true" == atbLtm["won"]
                    roster?.rank = (statsLtm["rank"] as Double).toInt()
                    roster?.teamId = (statsLtm["teamId"] as Double).toInt()
                    val rlsDataList = ((ltm["relationships"] as LinkedTreeMap<*, *>)["participants"] as LinkedTreeMap<*, *>)["data"] as List<*>
                    for (j in rlsDataList.indices) {
                        val rlsDataLtm = rlsDataList[j] as LinkedTreeMap<*, *>
                        val tmpParId = rlsDataLtm["id"] as String
                        if (matchDetail.participantDetailMap.containsKey(tmpParId)) {
                            matchDetail.participantDetailMap
                                .get(tmpParId)
                                ?.rosterId = tmpRosterId
                        } else {
                            val participantDetail = PubgParticipant(tmpParId)
                            participantDetail.rosterId = tmpRosterId
                            matchDetail.participantDetailMap.put(tmpParId, participantDetail)
                        }
                    }
                }
            }
        }
        return matchDetail
    }

    private fun parseParticipant(participant: PubgParticipant, treeMap: LinkedTreeMap<*, *>) {
        participant.dbnos = (treeMap["DBNOs"] as Double).toInt()
        participant.assists = (treeMap["assists"] as Double).toInt()
        participant.boosts = (treeMap["boosts"] as Double).toInt()
        participant.damageDealt = treeMap["damageDealt"] as Double
        participant.deathType = treeMap["deathType"] as String
        participant.headshotKills = (treeMap["headshotKills"] as Double).toInt()
        participant.heals = (treeMap["heals"] as Double).toInt()
        participant.killPlace = (treeMap["killPlace"] as Double).toInt()
        participant.killPoints = (treeMap["killPoints"] as Double).toInt()
        participant.killPointsDelta = treeMap["killPointsDelta"] as Double
        participant.killStreaks = (treeMap["killStreaks"] as Double).toInt()
        participant.kills = (treeMap["kills"] as Double).toInt()
        participant.lastKillPoints = (treeMap["lastKillPoints"] as Double).toInt()
        participant.lastWinPoints = (treeMap["lastWinPoints"] as Double).toInt()
        participant.longestKill = treeMap["longestKill"] as Double
        participant.mostDamage = (treeMap["mostDamage"] as Double).toInt()
        participant.name = treeMap["name"] as String
        participant.playerId = treeMap["playerId"] as String
        participant.revives = (treeMap["revives"] as Double).toInt()
        participant.rideDistance = treeMap["rideDistance"] as Double
        participant.roadKills = (treeMap["roadKills"] as Double).toInt()
        participant.teamKills = (treeMap["teamKills"] as Double).toInt()
        participant.timeSurvived = (treeMap["timeSurvived"] as Double).toInt()
        participant.vehicleDestroys = (treeMap["vehicleDestroys"] as Double).toInt()
        participant.walkDistance = treeMap["walkDistance"] as Double
        participant.weaponsAcquired = (treeMap["weaponsAcquired"] as Double).toInt()
        participant.winPlace = (treeMap["winPlace"] as Double).toInt()
        participant.winPoints = (treeMap["winPoints"] as Double).toInt()
        participant.winPointsDelta = treeMap["winPointsDelta"] as Double
    }

    companion object {
        // DateTimeFormatter is immutable and thread-safe.
        private val dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")

        // Gson is thread-safe
        private val gson = Gson()
    }
}
